pubsub/google/cloud/pubsub_v1/subscriber/_protocol/leaser.py

```diff
@@ -64,30 +64,32 @@ def bytes(self):
 
     def add(self, items):
         """Add messages to be managed by the leaser."""
-        for item in items:
-            # Add the ack ID to the set of managed ack IDs, and increment
-            # the size counter.
-            if item.ack_id not in self._leased_messages:
-                self._leased_messages[item.ack_id] = _LeasedMessage(
-                    added_time=time.time(), size=item.byte_size
-                )
-                self._bytes += item.byte_size
-            else:
-                _LOGGER.debug("Message %s is already lease managed", item.ack_id)
+        with self._operational_lock:
```

Comment:
From my read of the code, this lock is normally used to protect access to the `_thread` attrbute, created in `start` and destroyed in `stop`.  The worker thread runs `maintain_leases`, which copies the `_leased_messages` dict before iterating it.  What race are we trying to prevent here?  If `add`, `remove`, and `maintain_leases` need to synchronize access to `_leased_messages`, then perhaps we need a different lock for that?

Reply:
More generally:  we could improve auditability / comprehensibility of the M/T code by adding a comment for each lock created, documenting what attributes it is supposed to guard.

Reply:
I had the impression that "operational lock" is meant to synchronize "operations" on the leaser that can be initiated from external threads (e.g. `start` and `stop`), and used it to make the `add` / `remove` operations atomic. If this causes confusion, it might indeed make sense to use a separate lock, e.g. named `add_remove_lock`?

The motivation behind this is the observation in debug logs that `add()` and `remove()` operations can be intertwined if invoked from different threads at just the "right" moment. Both operations modify the internal `_leased_messages` dict and the `_bytes` counter.

The dict is probably fine, because `ack_id` is unique for every message, but I wouldn't say the same for `_bytes`.